types/index.d.ts

```diff
@@ -18,7 +23,12 @@ import {
  *
  * [Documentation](https://next-auth.js.org/configuration/options#options)
  */
-export interface NextAuthOptions {
+export interface NextAuthOptions<
+  TUser extends User = User,
+  TProfile extends AdapterProfile = AdapterProfile,
+  TSession extends AdapterSession = AdapterSession,
+  TVerificationRequest extends VerificationRequest = VerificationRequest
+> {
```

Comment:
I don't see why this was added here? Could you explain? If you want to override any of these, check out https://next-auth.js.org/getting-started/typescript#module-augmentation

Reply:
This is required so that we can demand conformance between the adapter and the events - by pulling values like `TUser` up to be a generic on the options object we enable Typescript to make sure that the User object our adapter expects is the same as the User object our events expect.

Reply:
They will be, if you simply reference the same `User` interface anywhere! Check out the module augmentation stuff! It is used instead of generics, because it works across submodules as well! (For example overriding the `Session` interface with module augmentation, you will get the same types on the `session()` callbacks first param, and the returned valued of `useSession()`, even though one comes from `next-auth` and the other in `next-auth/client`)

Reply:
Oh huh... Ok on second read through I am starting to actually grok what module augmentation is.

Ok if this is how you want to do things throughout then I'll just ditch the generics. I will just 2 cents that as a user of the library this is unexpected to me since I can't think of another library where you customize the base interfaces using module augmentation rather than just passing generics - even if you would have to pass them twice (server + client). I don't love it.

Reply:
We do document this here, so anyone using this library should expect it 😊:

https://next-auth.js.org/getting-started/typescript#module-augmentation

I first saw it on the `@material-ui` documentation page:

https://material-ui.com/guides/typescript/#customization-of-theme

it is very powerful, and works very neatly! you will only have to defined your interfaces in a single location.

Reply:
Yeah reading the linked TS handbook page from your docs page was what finally got it into my head. My instincts about this are still that it's too similar to hard-casting and scala-like `implicit` declarations - two things that I feel are very dangerous - for me to feel comfortable with it.  I'm going to have to sit on it for a day to really chew through why I don't like it, though. (I think it's something about the fact that next-auth is sandwiched between code you own on both sides.)

But I super disagree with the idea that there is less magic with this than generics - with generics it's very clear right in place what is being passed around. And if TS infers the wrong generic you can manually specify it. With this it feels much more like "ah yes you have told TS over in this one file that it should cast every single User coming from this module to be type MyUser and it will just silently and magically do that cast."

Reply:
BTW why did you leave the generics in Adapter at all? Seems like if you're going in on this, go whole hog.

Reply:
The thing is, those places where it is referenced, they HAVE TO match in ALL cases. There is just no place between for example what the session callaback returns and what the useSession hook gives you, where the value could change. Providing it in a single location and referencing it implicitly makes it impossible for the user to screw it up.

Reply:
@balazsorban44 hmm ok yeah that makes sense - generic typing with hooks is very very painful. I still don't love it but since React hasn't given us a better way I can't think of something that would make that more graceful.

Reply:
The `Adapters` interface will be used by third party, and I guess it's harder for us to forsee what users want to do. Maybe they have a database that won't much our interfaces 100%.

I'm also not quite there yet with the adapters/databases, but both @lluia and @kriood have seen most of the types/interfaces redesigns or even contributed and I got positive feedback from them.

If in the future we can actually simplify the Adapters interface as well, I'm all for it!

Reply:
Aha well I think there's the big difference in how we were approaching this - I've been coming at this from the adapter working upwards. I was trying to get the TUser in the options and callbacks to infer a type match to the user object in the adapter.

Basically I was thinking of next-auth as mediating with my ORM (Prisma in my case, but same with TypeORM) and the cookies/session for me, but that I would want to still use the underlying ORM object so that I could stick all sorts of extra values on my user object. So I was using the generics piping to run the specific user type that Prisma had generated in and out from the other server functions.

But you're quite right that that does nothing for the tidy typing on the client-side of things. Tbh, I'm not using the client side in practice other than plucking out the user's id from the session, so I had a bit of a blindspot there. I get better the motivation now and I gotta go to sleep. We'll tackle the adapters another time 😉

Reply:
Glad we came to an understanding! 

I'm gonna 😴 as well. 😁